File: Striver/DP/Q30.java
package Striver.DP;

public class Q30{
    // Minimum insertions/deletions to convert a string1 to string2

    // answer will be the length of string1 + length of string2 - 2*length of LCS

    // recursive solution: Time: O(2^n) Space: O(n)
    public static int lcs(String s1, String s2, int n, int m){
        if(n < 0 || m < 0) return 0;
        if(s1.charAt(n) == s2.charAt(m)) return 1 + lcs(s1, s2, n-1, m-1);
        else return Math.max(lcs(s1, s2, n-1, m), lcs(s1, s2, n, m-1));
    }

    public static int min_operations(String s1, String s2){
        int n = s1.length();
        int m = s2.length();
        int lcs = lcs(s1, s2, n-1, m-1);
        return n + m - 2*lcs;
    }

    // memoization: Time: O(n*m) Space: O(n*m)
    public static int lcs_memo(String s1, String s2, int n, int m, int[][] dp){
        if(n < 0 || m < 0) return 0;
        if(dp[n][m] != -1) return dp[n][m];
        if(s1.charAt(n) == s2.charAt(m)) return dp[n][m] = 1 + lcs_memo(s1, s2, n-1, m-1, dp);
        else return dp[n][m] = Math.max(lcs_memo(s1, s2, n-1, m, dp), lcs_memo(s1, s2, n, m-1, dp));
    }

    public static int min_operations_memo(String s1, String s2){
        int n = s1.length();
        int m = s2.length();
        int[][] dp = new int[n][m];
        for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) dp[i][j] = -1;
        int lcs = lcs_memo(s1, s2, n-1, m-1, dp);
        return n + m - 2*lcs;
    }

    // tabulation: Time: O(n*m) Space: O(n*m)
    public static int lcs_tab(String s1, String s2, int n, int m){
        int[][] dp = new int[n+1][m+1];
        for(int i = 0; i <= m; i++) dp[0][i] = 0;
        for(int i = 0; i <= n; i++) dp[i][0] = 0;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s1.charAt(i-1) == s2.charAt(j-1)) dp[i][j] = 1 + dp[i-1][j-1];
                else dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
            }
        }

        return dp[n][m];
    }

    public static int min_operations_tab(String s1, String s2){
        int n = s1.length();
        int m = s2.length();
        int lcs = lcs_tab(s1, s2, n, m);
        return n + m - 2*lcs;
    }

    // space optimization: Time: O(n*m) Space: O(n)
    public static int lcs_tab_opt(String s1, String s2, int n, int m){
        int[] dp = new int[m+1];
        int[] curr = new int[m+1];
        for(int i = 0; i <= m; i++) dp[i] = 0;

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s1.charAt(i-1) == s2.charAt(j-1)) curr[j] = 1 + dp[j-1];
                else curr[j] = Math.max(dp[j], curr[j-1]);
            }
            for(int j = 0; j <= m; j++) dp[j] = curr[j];
        }

        return dp[m];
    }

    public static int min_operations_tab_opt(String s1, String s2){
        int n = s1.length();
        int m = s2.length();
        int lcs = lcs_tab_opt(s1, s2, n, m);
        return n + m - 2*lcs;
    }

    public static void main(String[] args){
        String s1 = "heap";
        String s2 = "pea";
        System.out.println(min_operations(s1, s2));
        System.out.println(min_operations_memo(s1, s2));
        System.out.println(min_operations_tab(s1, s2));
        System.out.println(min_operations_tab_opt(s1, s2));
    }
}

File: Striver/DP/Q30.py
# minimum insertions/deletions to make string1 to string2
#ans will len(str1) + len(str2) - 2*lcs(str1, str2)

# recursive solution
def lcs(s1, s2, idx1, idx2):
    if idx1<0 or idx2<0:
        return 0
    
    if s1[idx1] == s2[idx2]:
        return 1 + lcs(s1, s2, idx1-1, idx2-1)
    
    return max(lcs(s1, s2, idx1-1, idx2), lcs(s1, s2, idx1, idx2-1))

def minOperations(s1, s2):
    return len(s1) + len(s2) - 2*lcs(s1, s2, len(s1)-1, len(s2)-1)


# memoization
def lcsMem(s1, s2, idx1, idx2, dp):
    if idx1<0 or idx2<0:
        return 0
    
    if dp[idx1][idx2] != -1:
        return dp[idx1][idx2]
    
    if s1[idx1] == s2[idx2]:
        dp[idx1][idx2] = 1 + lcsMem(s1, s2, idx1-1, idx2-1, dp)
        return dp[idx1][idx2]
    
    dp[idx1][idx2] = max(lcsMem(s1, s2, idx1-1, idx2, dp), lcsMem(s1, s2, idx1, idx2-1, dp))
    return dp[idx1][idx2]


def minOperationsMem(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [[-1] * m for _ in range(n)]
    return len(s1) + len(s2) - 2*lcsMem(s1, s2, n-1, m-1, dp)

# tabulation
def lcsTab(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [[0] * (n+1) for _ in range(m+1)]

    for i in range(0,m):
        dp[0][i] = 0

    for i in range(0,n):
        dp[i][0] = 0

    for i in range(1,n+1):
        for j in range(1,m+1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = 1 + dp[i-1][j-1]
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])
    
    return dp[n][m]

def minOperationsTab(s1, s2):
    return len(s1) + len(s2) - 2*lcsTab(s1, s2)

# space optimization
def lcsTabOpt(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [0] * (m+1)
    curr = [0] * (m+1)

    for i in range(1,m+1):
        dp[i] = 0

    for i in range(1,n+1):
        for j in range(1,m+1):
            if s1[i-1] == s2[j-1]:
                curr[j] = 1 + dp[j-1]
            else:
                curr[j] = max(dp[j], curr[j-1])
        dp = curr
        curr = [0] * (m+1)

    return dp[m]

def minOperationsTabOpt(s1, s2):
    return len(s1) + len(s2) - 2*lcsTabOpt(s1, s2)

s1 = "heap"
s2 = "pea"
print(minOperations(s1, s2))
print(minOperationsMem(s1, s2))
print(minOperationsTab(s1, s2))
print(minOperationsTabOpt(s1, s2))
